javalib/src/main/scala/java/util/Arrays.scala

```diff
@@ -82,64 +211,48 @@ object Arrays {
+    sortRangeImpl(a, fromIndex, toIndex)
 
-  @noinline def sort[T <: AnyRef](array: Array[T], comparator: Comparator[_ >: T]): Unit = {
-    implicit val ord = toOrdering(comparator).asInstanceOf[Ordering[AnyRef]]
-    sortAnyRefImpl(array.asInstanceOf[Array[AnyRef]])
-  }
+  @noinline def sort[T <: AnyRef](array: Array[T], comparator: Comparator[_ >: T]): Unit =
+    sortImpl(array)(new ComparatorArrayOps(comparator))
 
   @noinline def sort[T <: AnyRef](array: Array[T], fromIndex: Int, toIndex: Int,
       comparator: Comparator[_ >: T]): Unit = {
-    implicit val ord = toOrdering(comparator).asInstanceOf[Ordering[AnyRef]]
-    sortRangeAnyRefImpl(array.asInstanceOf[Array[AnyRef]], fromIndex, toIndex)
+    sortRangeImpl(array, fromIndex, toIndex)(new ComparatorArrayOps(comparator))
   }
 
   @inline
-  private def sortRangeImpl[@specialized T: ClassTag](
-      a: Array[T], fromIndex: Int, toIndex: Int)(implicit ord: Ordering[T]): Unit = {
+  private def sortRangeImpl[T](a: Array[T], fromIndex: Int, toIndex: Int)(
+      implicit ops: ArrayOps[T]): Unit = {
     checkRangeIndices(a, fromIndex, toIndex)
     stableMergeSort[T](a, fromIndex, toIndex)
   }
 
   @inline
-  private def sortRangeAnyRefImpl(a: Array[AnyRef], fromIndex: Int, toIndex: Int)(
-      implicit ord: Ordering[AnyRef]): Unit = {
-    checkRangeIndices(a, fromIndex, toIndex)
-    stableMergeSortAnyRef(a, fromIndex, toIndex)
-  }
-
-  @inline
-  private def sortImpl[@specialized T: ClassTag: Ordering](a: Array[T]): Unit =
-    stableMergeSort[T](a, 0, a.length)
-
-  @inline
-  private def sortAnyRefImpl(a: Array[AnyRef])(implicit ord: Ordering[AnyRef]): Unit =
-    stableMergeSortAnyRef(a, 0, a.length)
+  private def sortImpl[T](a: Array[T])(implicit ops: ArrayOps[T]): Unit =
+    stableMergeSort[T](a, 0, ops.length(a))
 
   private final val inPlaceSortThreshold = 16
 
-  /** Sort array `a` with merge sort and insertion sort,
-   *  using the Ordering on its elements.
-   */
+  /** Sort array `a` with merge sort and insertion sort. */
   @inline
-  private def stableMergeSort[@specialized K: ClassTag](a: Array[K],
-      start: Int, end: Int)(implicit ord: Ordering[K]): Unit = {
+  private def stableMergeSort[T](a: Array[T], start: Int, end: Int)(
+      implicit ops: ArrayOps[T]): Unit = {
     if (end - start > inPlaceSortThreshold)
-      stableSplitMerge(a, new Array[K](a.length), start, end)
+      stableSplitMerge(a, ops.createFrom(a, ops.length(a)), start, end)
     else
       insertionSort(a, start, end)
   }
 
   @noinline
-  private def stableSplitMerge[@specialized K](a: Array[K], temp: Array[K],
-      start: Int, end: Int)(implicit ord: Ordering[K]): Unit = {
+  private def stableSplitMerge[T](a: Array[T], temp: Array[T], start: Int,
+      end: Int)(implicit ops: ArrayOps[T]): Unit = {
     val length = end - start
     if (length > inPlaceSortThreshold) {
-      val middle = start + (length / 2)
+      val middle = start + (length >> 2)
```

Comment:
Could you explain this change or do you have reference I could read?

Reply:
Hum, that looks plain wrong, actually :sweat_smile: 

It should be `length >> 1`, which is a faster way of computing `length / 2` when you know that `length >= 0`. But there's no real reason to change that in this commit. It must have been a stray left from too many changes.

Reply:
Ok, I tried it locally and couldn't understand so not bad that I asked. :-)